File: src/types/businessRole.ts
// SPDX-License-Identifier: AGPL-3.0-only

// BusinessRole — 业务岗位（如 CEO / CTO / 产品经理）
// 与 AgentRole（抽象执行器）和 AgencyExpert（业务人才）正交：
// - AgentRole 回答「怎么干活」（executor / planner / researcher）
// - BusinessRole 回答「在组织里担什么责」
// - AgencyExpert 回答「具体技能是什么」
// AgentProfile 通过 business_role_id + expert_id 将三者融合。

/**
 * 业务岗位来源
 * - builtin: 内置预设
 * - custom: 用户自定义
 */
export type BusinessRoleSource = "builtin" | "custom";

/** 资历等级（与 AgencyExpert 保持一致） */
export type SeniorityLevel = "junior" | "mid" | "senior" | "expert";

/**
 * 业务岗位 DTO，字段对齐后端 `BusinessRoleDto`
 * （`src-tauri/crates/harness/src/repo_dtos.rs`）。
 *
 * 注意：后端 responsibilities / decision_authority / managed_expert_ids /
 * required_certifications / active_domains 均以 JSON 字符串形式存储，
 * 前端在 store 层做 parse / stringify 转换，UI 层直接使用结构化数组/对象。
 */
export interface BusinessRole {
  id: string;
  name: string;
  description: string | null;
  /** 岗位核心职责列表（已 parse） */
  responsibilities: string[];
  /** 决策权限边界（已 parse，结构由业务自定义） */
  decisionAuthority: Record<string, unknown> | null;
  /** 汇报对象 ID（business_roles.id 自引用） */
  reportsTo: string | null;
  /** 下属专家 ID 列表（已 parse） */
  managedExpertIds: string[];
  /** 准入条件列表（已 parse） */
  requiredCertifications: string[];
  /** 激活业务域列表（已 parse，如 ["core", "invest"]） */
  activeDomains: string[];
  /** 岗位系统提示词（4 层 prompt 拼接的最外层） */
  systemPrompt: string;
  icon: string | null;
  color: string | null;
  source: BusinessRoleSource;
  sortOrder: number;
  isEnabled: boolean;
  createdAt: number;
  updatedAt: number;
}

/**
 * 创建/更新业务岗位的输入。
 * 严格对齐后端 `SaveBusinessRoleInput`
 * （`src-tauri/src/commands/business_role.rs`）。
 *
 * 语义为 upsert：前端创建时需自行生成 uuid 作为 id；
 * 更新时传入已有 id。后端创建时 is_enabled 默认为 true，
 * 目前未提供通过 save 命令修改 is_enabled 的入口。
 */
export interface SaveBusinessRoleInput {
  /** 必填，前端创建时用 crypto.randomUUID() 生成 */
  id: string;
  name: string;
  description?: string | null;
  responsibilities?: string[];
  /**
   * 决策权限边界，后端以 JSON 字符串存储。
   * 前端 store 层负责将对象 JSON.stringify 后传入；
   * 也可直接传字符串。
   */
  decisionAuthority?: string | null;
  reportsTo?: string | null;
  managedExpertIds?: string[];
  requiredCertifications?: string[];
  activeDomains?: string[];
  /** 必填，岗位系统提示词（4 层 prompt 拼接的最外层） */
  systemPrompt: string;
  icon?: string | null;
  color?: string | null;
  source?: BusinessRoleSource;
  sortOrder?: number;
}

File: src-tauri/src/commands/business_role.rs
// SPDX-License-Identifier: AGPL-3.0-only

//! 业务岗位命令 —— 对应现实业务岗位（CEO/CTO/产品经理 等）。
//!
//! 与 `agent_role` 命令（抽象执行器类型）区别：
//! - `agent_role`：executor/planner/researcher 等执行器类型，描述「怎么干活」
//! - `business_role`：CEO/CTO/产品经理 等业务岗位，描述「在组织里担什么责」
//!
//! 二者通过 `agent_profiles.business_role_id` + `agent_profiles.agent_role` 同时关联。

use agent_macro::agent_command;

use crate::AppState;
use axagent_dao::repo::business_role;
use axagent_harness::repo_dtos::BusinessRoleDto;
use serde::Deserialize;
use tauri::State;

/// 列出所有业务岗位
#[agent_command(domain = "agent", safety = Safe, call_mode = StateInput, description = "列出所有业务岗位")]
#[tauri::command]
pub async fn list_business_roles(
    app_state: State<'_, AppState>,
    source: Option<String>,
) -> Result<Vec<BusinessRoleDto>, String> {
    business_role::list_business_roles(app_state.harness.db(), source.as_deref()).await.map_err(
        |e| {
            String::from(crate::commands::error::ErrorResponse::from_error(
                e,
                crate::commands::error::ErrorCategory::Unrecoverable,
            ))
        },
    )
}

/// 查询业务岗位树（按 sort_order 排序，前端按 reports_to 自行构造树）
#[agent_command(domain = "agent", safety = Safe, call_mode = StateOnly, description = "查询业务岗位树")]
#[tauri::command]
pub async fn list_business_role_tree(
    app_state: State<'_, AppState>,
) -> Result<Vec<BusinessRoleDto>, String> {
    business_role::list_business_role_tree(app_state.harness.db()).await.map_err(|e| {
        String::from(crate::commands::error::ErrorResponse::from_error(
            e,
            crate::commands::error::ErrorCategory::Unrecoverable,
        ))
    })
}

/// 获取单个业务岗位
#[agent_command(domain = "agent", safety = Safe, call_mode = StateInput, description = "获取单个业务岗位")]
#[tauri::command]
pub async fn get_business_role(
    app_state: State<'_, AppState>,
    id: String,
) -> Result<BusinessRoleDto, String> {
    business_role::get_business_role(app_state.harness.db(), &id).await.map_err(|e| {
        String::from(crate::commands::error::ErrorResponse::from_error(
            e,
            crate::commands::error::ErrorCategory::Unrecoverable,
        ))
    })
}

/// 创建/更新业务岗位的输入参数
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveBusinessRoleInput {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub responsibilities: Option<Vec<String>>,
    pub decision_authority: Option<String>,
    pub reports_to: Option<String>,
    pub managed_expert_ids: Option<Vec<String>>,
    pub required_certifications: Option<Vec<String>>,
    pub active_domains: Option<Vec<String>>,
    pub system_prompt: String,
    pub icon: Option<String>,
    pub color: Option<String>,
    pub source: Option<String>,
    pub sort_order: Option<i32>,
}

/// 保存（创建/更新）业务岗位
#[agent_command(domain = "agent", safety = Caution, call_mode = StateInput, description = "保存业务岗位")]
#[tauri::command]
pub async fn save_business_role(
    app_state: State<'_, AppState>,
    input: SaveBusinessRoleInput,
) -> Result<BusinessRoleDto, String> {
    let source = input.source.as_deref().unwrap_or("custom");
    let sort_order = input.sort_order.unwrap_or(0);
    business_role::upsert_business_role(
        app_state.harness.db(),
        &input.id,
        &input.name,
        input.description.as_deref(),
        input.responsibilities.as_deref(),
        input.decision_authority.as_deref(),
        input.reports_to.as_deref(),
        input.managed_expert_ids.as_deref(),
        input.required_certifications.as_deref(),
        input.active_domains.as_deref(),
        &input.system_prompt,
        input.icon.as_deref(),
        input.color.as_deref(),
        source,
        sort_order,
    )
    .await
    .map_err(|e| {
        String::from(crate::commands::error::ErrorResponse::from_error(
            e,
            crate::commands::error::ErrorCategory::Unrecoverable,
        ))
    })
}

/// 删除业务岗位（builtin 不可删除；有下属岗位时拒绝删除）
#[agent_command(domain = "agent", safety = Dangerous, call_mode = StateInput, description = "删除业务岗位")]
#[tauri::command]
pub async fn delete_business_role(
    app_state: State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    let role =
        business_role::get_business_role(app_state.harness.db(), &id).await.map_err(|e| {
            String::from(crate::commands::error::ErrorResponse::from_error(
                e,
                crate::commands::error::ErrorCategory::Unrecoverable,
            ))
        })?;

    if role.source == "builtin" {
        return Err("内置业务岗位不可删除".to_string());
    }

    business_role::delete_business_role(app_state.harness.db(), &id).await.map_err(|e| {
        String::from(crate::commands::error::ErrorResponse::from_error(
            e,
            crate::commands::error::ErrorCategory::Unrecoverable,
        ))
    })
}
